pkg/scheduler/controller.go

```diff
@@ -0,0 +1,249 @@
+	})
+
+	locationAccess.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: c.enqueueLocation,
+		UpdateFunc: func(old, obj interface{}) {
+			oldLoc := old.(*schedulingv1alpha1.Location)
+			newLoc := obj.(*schedulingv1alpha1.Location)
+			if !apiequality.Semantic.DeepEqual(oldLoc.Spec, newLoc.Spec) || !apiequality.Semantic.DeepEqual(oldLoc.Labels, newLoc.Labels) {
+				c.enqueueLocation(obj)
+			}
+		},
+		DeleteFunc: c.enqueueLocation,
+	})
+
+	syncTargetAccess.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: c.enqueueSyncTarget,
+		UpdateFunc: func(old, obj interface{}) {
+			oldST := old.(*workloadv1alpha1.SyncTarget)
+			newST := obj.(*workloadv1alpha1.SyncTarget)
+			if !apiequality.Semantic.DeepEqual(oldST.Spec, newST.Spec) || !apiequality.Semantic.DeepEqual(oldST.Labels, newST.Labels) {
+				c.enqueueSyncTarget((obj))
+			}
+		},
+		DeleteFunc: c.enqueueSyncTarget,
+	})
+
+	return c, nil
+}
+
+func (c *controller) enqueueEdgePlacement(obj interface{}) {
+	key, err := kcpcache.DeletionHandlingMetaClusterNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	klog.FromContext(c.context).V(2).Info("queueing EdgePlacement", "key", key)
+	c.queue.Add(
+		queueItem{
+			triggeringKind: triggeringKindEdgePlacement,
+			key:            key,
+		},
+	)
+}
+
+func (c *controller) enqueueLocation(obj interface{}) {
+	key, err := kcpcache.DeletionHandlingMetaClusterNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	klog.FromContext(c.context).V(2).Info("queueing Location", "key", key)
+	c.queue.Add(
+		queueItem{
+			triggeringKind: triggeringKindLocation,
+			key:            key,
+		},
+	)
+}
+
+func (c *controller) enqueueSyncTarget(obj interface{}) {
+	key, err := kcpcache.DeletionHandlingMetaClusterNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	klog.FromContext(c.context).V(2).Info("queueing SyncTarget", "key", key)
+	c.queue.Add(
+		queueItem{
+			triggeringKind: triggeringKindSyncTarget,
+			key:            key,
+		},
+	)
+}
+
+// Run starts the controller, which stops when c.context.Done() is closed.
+func (c *controller) Run(numThreads int) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	logger := klog.FromContext(c.context)
+	logger.Info("starting controller")
+	defer logger.Info("shutting down controller")
+
+	for i := 0; i < numThreads; i++ {
+		go wait.UntilWithContext(c.context, c.startWorker, time.Second)
+	}
+
+	<-c.context.Done()
+}
+
+func (c *controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	i, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	item := i.(queueItem)
+	key := item.key
+
+	klog.FromContext(ctx).V(1).Info("processing key", "triggeringKind", item.triggeringKind, "key", key)
```

Comment:
What do you think of the following pattern, which I have seen other controllers use?

At this point in the logic, first make a logger that logs more stuff and a corresponding context:
```
	ctx = klog.NewContext(ctx, klog.FromContext(ctx).WithValues("triggeringKind", item.triggeringKind, "key", key))
```

From that point on, the information already added to that logger does not have to be explicitly included in any downstream logging statements.  So, for example, we would see the following.

```
	klog.FromContext(ctx).V(1).Info("processing queueItem")
```

Reply:
Thanks for the details! Followed the pattern by https://github.com/kcp-dev/edge-mc/pull/206/commits/5a5cd2a9e9c74a3fd56d2120f37985877aedd50b.